### ISSUE TITLE
Push Constant verification

### ISSUE THREAD

**Issue by nicokoch** [vulkano-rs/vulkano#756]
I noticed there are no bounds on the push constant parameter in `AutoCommandBufferBuilder::draw`.

Is it possible to verify the the data with regards to the shader layout (or pipeline layout) or are there technical limitations?
I imagine something similar to `impl_vertex!`, but for push constants.

Can the current implementation lead to invalid API usage?

**Comment by tomaka** [comment]
> Can the current implementation lead to invalid API usage?

No. [The code](https://github.com/vulkano-rs/vulkano/blob/06a0b3715ff235ce78c4f50e580134467be44ef0/vulkano/src/command_buffer/auto.rs#L1073-L1091) only uploads the ranges corresponding to the ranges of the pipeline layout, and the pipeline layout itself is verified.

> I imagine something similar to impl_vertex!, but for push constants.

That would be too complex, and if we do it it would lead to a large overhead I think.
Contrary to vertex attributes, the push constants can contain structs, must respect alignment, and so on.

**Comment by nicokoch** [comment]
You're right I assume.

So in order to set my push constants I can use a `#repr(c)` struct in rust code that resembles the `layout(push_constant) uniform` in my shader.
It's just that vulkano doesn't check if the structs match.
Correct?

**Comment by tomaka** [comment]
> Correct?

Yes.

Also note that vulkano-shaders will generate a struct a compile-time that you can use.

**Comment by knappador** [comment]
Just ran into a bug created by me defining a shader that uses push constants but forgetting to pass constants I had created to the dispatch command.  Garbage eventually happened and showed me that I had not passed them to the command buffer builder.  Making a note to investigate if this behavior is not undefined (in my case it lead to garbage reads) and also to come up with some compile time restriction to block it.
